File: LeetCode/palindrome-partitioning-ii.c
int min(int a,int b){
   if(a<b){
       return a;
   } 
   return b;
}
int minCut(char* s) {
    int len = strlen(s);
        if(len==0){
            return 0;
        }
        int p[len][len];
        int c[len][len];
        
        for(int i=0;i<len;i++){
            p[i][i]=1;
            c[i][i]=0;
        }
        for(int l=2;l<=len;l++){
            for(int i=0;i<len-l+1;i++){
                int j=l+i-1;
                if(l==2){
                   if(s[i]==s[j]){
                       p[i][j]=1;
                   }
                }else{
                    if((s[i]==s[j]) && p[i+1][j-1]==1){
                       p[i][j]=1;
                   }
                }
                
                if(p[i][j]==1){
                    c[i][j]=0;
                }else{
                    c[i][j] = INT_MAX;
                    for(int k=i;k<=j-1;k++){
                       c[i][j]=min(c[i][j],c[i][k]+c[k+1][j]+1); 
                    }
                }
            }
        }
        return c[0][len-1];
}

File: LeetCode/palindrome-partitioning-ii.java
public class Solution {
    public int minCut(String s) {
        int len = s.length();
        if(len==0 || len==1){
            return 0;
        }
		int[] cuts = new int[len+1];
		for(int i=0;i<=len;i++){
			cuts[i]=i-1;
		}
		for(int j=0;j<len;j++){
			for(int i=0;(j-i>=0 && i+j<len) && s.charAt(j-i)==s.charAt(i+j);i++){
				cuts[i+j+1]=Math.min(cuts[i+j+1], cuts[j-i]+1);
						
			}
			for(int i=1;(j-i+1>=0 && i+j<len) && s.charAt(j-i+1)==s.charAt(i+j);i++){
				cuts[i+j+1]=Math.min(cuts[j-i+1]+1,cuts[i+j+1]);
			}
			
		}
		return cuts[len];
    }
}